contracts/sg-ics721/src/testing/integration_tests.rs

```diff
@@ -0,0 +1,1426 @@
+    // need to wrap method in function for testing
+    fn ibc_reply(deps: DepsMut, env: Env, reply: Reply) -> Result<Response, ContractError> {
+        SgIcs721Contract::default()
+            .reply(deps, env, reply)
+            .and_then(|_| Ok(Response::default()))
+    }
+
+    let contract = ContractWrapper::new(execute, instantiate, query)
+        .with_migrate(migrate)
+        .with_reply(ibc_reply);
+    Box::new(contract)
+}
+
+fn proxy_contract() -> Box<dyn Contract<Empty>> {
+    let execute_fn = cw721_rate_limited_proxy::contract::execute::<Empty>;
+    let instatiate_fn = cw721_rate_limited_proxy::contract::instantiate::<Empty>;
+    let contract = ContractWrapper::new(
+        execute_fn,
+        instatiate_fn,
+        cw721_rate_limited_proxy::contract::query,
+    );
+    Box::new(contract)
+}
+
+#[test]
+fn test_instantiate() {
+    let mut test = Test::new(false, None, sg721_base_contract());
+
+    // check stores are properly initialized
+    let cw721_id = test.query_cw721_id();
+    assert_eq!(cw721_id, test.cw721_id);
+    let nft_contracts: Vec<(String, Addr)> = test.query_nft_contracts();
+    assert_eq!(nft_contracts, Vec::<(String, Addr)>::new());
+    let outgoing_channels = test.query_outgoing_channels();
+    assert_eq!(outgoing_channels, []);
+    let incoming_channels = test.query_incoming_channels();
+    assert_eq!(incoming_channels, []);
+}
+
+#[test]
+fn test_do_instantiate_and_mint_weird_data() {
+    let mut test = Test::new(false, None, sg721_base_contract());
+
+    test.app
+        .execute_contract(
+            test.ics721.clone(),
+            test.ics721,
+            &ExecuteMsg::Callback(CallbackMsg::CreateVouchers {
+                receiver: "mr-t".to_string(),
+                create: VoucherCreation {
+                    class: Class {
+                        id: ClassId::new("bad kids"),
+                        uri: None,
+                        data: Some(
+                            to_binary(&ClassData {
+                                owner: Some(OWNER_SOURCE_CHAIN.to_string()),
+                                contract_info: Default::default(),
+                                name: "name".to_string(),
+                                symbol: "symbol".to_string(),
+                                num_tokens: 1,
+                            })
+                            .unwrap(),
+                        ),
+                    },
+                    tokens: vec![Token {
+                        id: TokenId::new("1"),
+                        // IMPORTANT: unlike cw721-base, for sg721-base empty URI string is NOT allowed
+                        uri: Some("arkprotocol".to_string()),
+                        data: None,
+                    }],
+                },
+            }),
+            &[],
+        )
+        .unwrap();
+}
+
+#[test]
+fn test_do_instantiate_and_mint() {
+    // test case: instantiate cw721 with no ClassData
```

Comment:
Same test, but this time for sg-ics721 contract, checking whether creator is correctly set, based on source owner/minter address (provided by cw721-base).